okhttp-tests/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java

```diff
@@ -773,6 +773,47 @@ private static OkHttpClient buildHttp2Client() {
     assertEquals(1, server.takeRequest().getSequenceNumber()); // Reused connection.
   }
 
+  @Test public void recoverFromCancelReusesConnection() throws Exception {
+    // Set up two responses
+    server.enqueue(new MockResponse()
+            .setBodyDelay(10, TimeUnit.SECONDS)
+            .setBody("abc"));
+    server.enqueue(new MockResponse()
+            .setBody("def"));
+
+    client = client.newBuilder()
+            .dns(new DoubleInetAddressDns())
+            .build();
+
+    Call call1 = client.newCall(new Request.Builder()
+            .url(server.url("/"))
+            .build());
+    // Use async handler so that we can cancel before receiving the body
+    // We don't really care about the response since we'll cancel immediately,
+    // so we don't do anything on callback
+    Callback callback = new Callback() {
```

Comment:
You should be able to use execute, and just not read the body.  It should make the test simpler.

Reply:
Using execute causes SocketTimeoutExceptions. Does execute wait to read the body before returning?